File: Java/src/rip/rip.java
package rip;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.InputMismatchException;
import java.util.Scanner;

public class rip {
    public static final int INFINITY = 1024;

    public static boolean ifChanged = true;

    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("src/rip/test.txt");
        RoutingInfo[][] routingTable = initRountingTable(f);
        int nodeNumber = routingTable.length;
        while (ifChanged) {
            ifChanged = false;
            sendRoutingTable(routingTable);
            for (RoutingInfo[] routingInfos : routingTable) {
                for (int j = 0; j < nodeNumber; j++) {
                    if (routingInfos[j].getCost() == INFINITY) {
                        System.out.print("* ");
                    } else {
                        System.out.print(routingInfos[j].getCost() + " ");
                    }
                }
                System.out.println();
            }
            System.out.println("------------------------------------");
        }
    }

    public static boolean isAdjacent(RoutingInfo[][] routingTable, int nodeNum1, int nodeNum2) {
        for (int i = 0; i < routingTable.length; i++) {
            if (routingTable[nodeNum1][i].getNextHop() == nodeNum2) {
                return true;
            }
        }
        return false;
    }

    public static void receiveRoutingTable(RoutingInfo[][] routingTable, int senderNum, int receiverNum, int[] senderInfo) {
        for (int i = 0; i < routingTable.length; i++) {
            if (routingTable[receiverNum][i].getCost() > routingTable[receiverNum][senderNum].getCost() + senderInfo[i]) {
                routingTable[receiverNum][i].setCost(routingTable[receiverNum][senderNum].getCost() + senderInfo[i]);
                routingTable[receiverNum][i].setNextHop(senderNum);
                ifChanged = true;
            }
        }
    }

    public static void sendRoutingTable(RoutingInfo[][] routingTable) {
        int nodeNumber = routingTable.length;
        int[][] senderInfos = new int[nodeNumber][nodeNumber];
        for (int i = 0; i < nodeNumber; i++) {
            for (int j = 0; j < nodeNumber; j++) {
                senderInfos[i][j] = routingTable[i][j].getCost();
            }
        }
        for (int i = 0; i < nodeNumber; i++) {
            for (int j = 0; j < nodeNumber; j++) {
                if (isAdjacent(routingTable, i, j)) {
                    receiveRoutingTable(routingTable, i, j, senderInfos[i]);
                }
            }
        }
    }

    public static RoutingInfo[][] initRountingTable(File f) throws FileNotFoundException {
        Scanner sc = new Scanner(f);
        int nodeNumber = sc.nextInt();
        RoutingInfo[][] routingTable = new RoutingInfo[nodeNumber][nodeNumber];
        for (int i = 0; i < nodeNumber; i++) {
            for (int j = 0; j < nodeNumber; j++) {
                routingTable[i][j] = new RoutingInfo();
            }
        }
        int token;
        String buffer;
        while (sc.hasNext()) {
            for (RoutingInfo[] routingInfos : routingTable) {
                for (int j = 0; j < routingInfos.length; j++) {
                    try {
                        token = sc.nextInt();
                        routingInfos[j].setCost(token);
                    } catch (InputMismatchException e) {
                        buffer = sc.next();
                        if (buffer.equals("i")) {
                            routingInfos[j].setCost(INFINITY);
                        }
                    }
                    routingInfos[j].setDestination(j);
                    routingInfos[j].setNextHop(routingInfos[j].getCost() == INFINITY || routingInfos[j].getCost() == 0 ? routingInfos.length : j);
                }
            }
        }
        return routingTable;
    }
}

File: Go/rip/rip.go
package main

import (
	"fmt"
	"io"
	"os"
	"strconv"
)

const INFINITY int = 1024

type RoutingInfo struct {
	destination int
	cost        int
	nextHop     int
}

var ifChanged = true

func initRoutingTable(f *os.File) [][]RoutingInfo {
	var nodeNum int
	_, err := fmt.Fscanf(f, "%d", &nodeNum)
	if err != nil {
		fmt.Printf("read file false")
		os.Exit(1)
	}
	routingTable := make([][]RoutingInfo, nodeNum)
	for i := 0; i < nodeNum; i++ {
		for j := 0; j < nodeNum; j++ {
			var routingInfo RoutingInfo
			routingTable[i] = append(routingTable[i], routingInfo)
		}
	}
	var token string
	_, err = fmt.Fscan(f, &token)
	if err != nil {
		fmt.Printf("read file false")
		os.Exit(1)
	}
	for i := 0; i < nodeNum; i++ {
		for j := 0; j < nodeNum; j++ {
			temp, err := strconv.Atoi(token)
			if err == nil {
				routingTable[i][j].cost = temp
			} else {
				if token == "i" {
					routingTable[i][j].cost = INFINITY
				}
			}
			routingTable[i][j].destination = j
			if routingTable[i][j].cost == INFINITY || routingTable[i][j].cost == 0 {
				routingTable[i][j].nextHop = len(routingTable)
			} else {
				routingTable[i][j].nextHop = j
			}
			_, err = fmt.Fscan(f, &token)
			if err == io.EOF {
			} else if err != nil {
				fmt.Printf("read file false")
				os.Exit(1)
			}
		}
	}
	return routingTable
}

func isAdjacent(routingTable [][]RoutingInfo, nodeNum1 int, nodeNum2 int) bool {
	for i := 0; i < len(routingTable); i++ {
		if routingTable[nodeNum1][i].nextHop == nodeNum2 {
			return true
		}
	}
	return false
}

func receiveRoutingTable(routingTable [][]RoutingInfo, senderNum int, receicerNum int, senderInfo []int) {
	for i := 0; i < len(routingTable); i++ {
		if routingTable[receicerNum][i].cost > routingTable[receicerNum][senderNum].cost+senderInfo[i] {
			routingTable[receicerNum][i].cost = routingTable[receicerNum][senderNum].cost + senderInfo[i]
			routingTable[receicerNum][i].nextHop = senderNum
			ifChanged = true
		}
	}
}

func sendRoutingTable(routingTable [][]RoutingInfo) {
	nodeNum := len(routingTable)
	senderInfos := make([][]int, nodeNum)
	for i := 0; i < nodeNum; i++ {
		for j := 0; j < nodeNum; j++ {
			senderInfos[i] = append(senderInfos[i], routingTable[i][j].cost)
		}
	}
	for i := 0; i < nodeNum; i++ {
		for j := 0; j < nodeNum; j++ {
			if isAdjacent(routingTable, i, j) {
				receiveRoutingTable(routingTable, i, j, senderInfos[i])
			}
		}
	}
}

func main() {
	f, err := os.Open("rip/test.txt")
	if err != nil {
		fmt.Printf("file error")
		os.Exit(1)
	}
	routingTable := initRoutingTable(f)
	nodeNum := len(routingTable)
	for ifChanged {
		ifChanged = false
		sendRoutingTable(routingTable)
		for i := 0; i < nodeNum; i++ {
			for j := 0; j < nodeNum; j++ {
				if routingTable[i][j].cost == INFINITY {
					fmt.Printf("%s ", "*")
				} else {
					fmt.Printf("%d ", routingTable[i][j].cost)
				}
			}
			fmt.Println()
		}
		fmt.Println("---------------------------------------")
	}
}
